From ce2615164a4801c0cd74e09e9a574f0efd284770 Mon Sep 17 00:00:00 2001
From: Rajendra Dendukuri <rajendra.dendukuri@broadcom.com>
Date: Wed, 29 May 2019 10:25:29 -0400
Subject: [PATCH] Policy to wait for IPv6 link local address to be available

Add a ifupdown2 policy attribute to allow user to tune the wait time to be
allowed before the dhclient can be started on an IPv6 enabled interface.
---
 ifupdown2/addons/dhcp.py | 21 +++++++++++++++++----
 1 file changed, 17 insertions(+), 4 deletions(-)

diff --git a/ifupdown2/addons/dhcp.py b/ifupdown2/addons/dhcp.py
index 09251a9..c3c125c 100644
--- a/ifupdown2/addons/dhcp.py
+++ b/ifupdown2/addons/dhcp.py
@@ -61,6 +61,14 @@ class dhcp(moduleBase):
             dhcp_wait = policymanager.policymanager_api.get_attr_default(
                 module_name=self.__class__.__name__, attr='dhcp-wait')
             wait = not str(dhcp_wait).lower() == "no"
+            inet6_ll_wait = policymanager.policymanager_api.get_iface_default(module_name=self.__class__.__name__, \
+                ifname=ifaceobj.name, attr='inet6-ll-wait')
+            try:
+                ll_wait_time = int(inet6_ll_wait)
+            except:
+                ll_wait_time = 10
+                pass
+
             vrf = ifaceobj.get_attr_value_first('vrf')
             if (vrf and self.vrf_exec_cmd_prefix and
                 self.ipcmd.link_exists(vrf)):
@@ -100,10 +108,13 @@ class dhcp(moduleBase):
                             pass
                     #add delay before starting IPv6 dhclient to
                     #make sure the configured interface/link is up.
-                    time.sleep(2)
-                    timeout = 10
+                    if ll_wait_time:
+                       timeout = ll_wait_time
+                       time.sleep(1)
+                    else:
+                       timeout = ll_wait_time+1
+
                     while timeout:
-                        timeout -= 2
                         addr_output = utils.exec_command('%s -6 addr show %s'
                                                          %(utils.ip_cmd, ifaceobj.name))
                         r = re.search('inet6 .* scope link', addr_output)
@@ -112,7 +123,9 @@ class dhcp(moduleBase):
                                                     wait=wait,
                                                     cmd_prefix=dhclient_cmd_prefix)
                             return
-                        time.sleep(2)
+                        timeout -= 1
+                        if timeout:
+                            time.sleep(1)

         except Exception, e:
             self.log_error(str(e), ifaceobj)
--
2.7.4

